code/game/machinery/atm_ret_field.dm

```diff
@@ -0,0 +1,246 @@
+			to_chat(user, "<span class='notice'>You fully disassemble \the [src]. There were no salvageable parts.</span>")
+			qdel(src)
+		return
+
+/obj/machinery/atmospheric_field_generator/perma/Initialize()
+	. = ..()
+	generate_field()
+
+/obj/machinery/atmospheric_field_generator/update_icon()
+	if(stat & BROKEN)
+		icon_state = "arfg_broken"
+	else if(hatch_open && wires_intact)
+		icon_state = "arfg_open_wires"
+	else if(hatch_open && !wires_intact)
+		icon_state = "arfg_open_wirescut"
+	else if(isactive)
+		icon_state = "arfg_on"
+	else
+		icon_state = "arfg_off"
+
+/obj/machinery/atmospheric_field_generator/power_change()
+	var/oldstat
+	. = ..()
+	if(!(stat & NOPOWER))
+		ispowered = 1
+		update_icon()
+		if(alwaysactive || wasactive)	//reboot our field if we were on or are supposed to be always-on
+			generate_field()
+	if(stat != oldstat && isactive && (stat & NOPOWER))
+		ispowered = 0
+		disable_field()
+		update_icon()
+
+/obj/machinery/atmospheric_field_generator/emp_act()
+	. = ..()
+	disable_field() //shutting dowwwwwwn
+	if(alwaysactive || wasactive) //reboot after a short delay if we were online before
+		spawn(rand(reboot_delay_min,reboot_delay_max))
+			generate_field()
+
+/obj/machinery/atmospheric_field_generator/ex_act(severity)
+	switch(severity)
+		if(1)
+			disable_field()
+			qdel(src)
+			return
+		if(2)
+			stat |= BROKEN
+			update_icon()
+			src.visible_message("The ARF-G cracks and shatters!","You hear an uncomfortable metallic crunch.")
+			disable_field()
+		if(3)
+			emp_act()
+	return
+
+/obj/machinery/atmospheric_field_generator/proc/generate_field()
+	if(!ispowered || hatch_open || !wires_intact || isactive) //if it's not powered, the hatch is open, the wires are busted, or it's already on, don't do anything
+		return
+	else
+		isactive = 1
+		icon_state = "arfg_on"
+		new field_type (src.loc)
+		src.visible_message("<span class='warning'>The ARF-G crackles to life!</span>","<span class='warning'>You hear an ARF-G coming online!</span>")
+		update_use_power(USE_POWER_ACTIVE)
+	return
+
+/obj/machinery/atmospheric_field_generator/proc/disable_field()
+	if(isactive)
+		icon_state = "arfg_off"
+		for(var/obj/structure/atmospheric_retention_field/F in loc)
+			qdel(F)
+		src.visible_message("The ARF-G shuts down with a low hum.","You hear an ARF-G powering down.")
+		update_use_power(USE_POWER_IDLE)
+		isactive = 0
+	return
+
+/obj/machinery/atmospheric_field_generator/Initialize()
+	. = ..()
+	//Delete ourselves if we find extra mapped in arfgs
+	for(var/obj/machinery/atmospheric_field_generator/F in loc)
+		if(F != src)
+			log_debug("Duplicate ARFGS at [x],[y],[z]")
+			return INITIALIZE_HINT_QDEL
+
+	var/area/A = get_area(src)
+	ASSERT(istype(A))
+
+	LAZYADD(A.all_arfgs, src)
+	areas_added = list(A)
+
+	for(var/direction in cardinal)
+		A = get_area(get_step(src,direction))
+		if(istype(A) && !(A in areas_added))
+			LAZYADD(A.all_arfgs, src)
+			areas_added += A
```

Comment:
Should this be a `LAZYADD()` too? Do we not care about duplicates in the list? (Suppose it's on a corner)